File: generate_did.py
try:
    import sys
    import json
    if sys.stdout and hasattr(sys.stdout, 'reconfigure'):
        sys.stdout.reconfigure(encoding='utf-8')
    from cryptography.hazmat.primitives.asymmetric import ed25519
    import base58
except ImportError:
    print("❌ Thiếu thư viện! Vui lòng cài đặt trước:")
    print("   pip install cryptography base58")
    sys.exit(1)

def generate_did_key():
    # 1. Tạo Ed25519 Private Key
    private_key = ed25519.Ed25519PrivateKey.generate()
    public_key = private_key.public_key()
    
    # 2. Lấy raw bytes (32 bytes mỗi loại)
    pub_bytes = public_key.public_bytes_raw()
    priv_bytes = private_key.private_bytes_raw()
    
    # 3. Multicodec prefix cho Ed25519 public key = 0xed01 (2 bytes)
    multicodec_prefix = b'\xed\x01'
    multicodec_pub = multicodec_prefix + pub_bytes
    
    # 4. Base58btc encoding với tiền tố multibase 'z'
    did_key = "did:key:z" + base58.b58encode(multicodec_pub).decode('ascii')
    
    return {
        "did_key": did_key,
        "public_key_hex": pub_bytes.hex(),
        "private_key_hex": priv_bytes.hex()
    }

def sign_payload(private_key_hex: str, message: str) -> str:
    priv_bytes = bytes.fromhex(private_key_hex)
    private_key = ed25519.Ed25519PrivateKey.from_private_bytes(priv_bytes)
    signature = private_key.sign(message.encode('utf-8'))
    return signature.hex()

if __name__ == "__main__":
    key_pair = generate_did_key()
    print("==================================================")
    print("🔑 TECHNOCORE DID KEY GENERATOR (Ed25519)")
    print("==================================================")
    print(f"DID Key         : {key_pair['did_key']}")
    print(f"Public Key (Hex): {key_pair['public_key_hex']}")
    print(f"Private Key(Hex): {key_pair['private_key_hex']}")
    print("==================================================")
    print("⚠️  LƯU Ý BẢO MẬT: Giữ bí mật Private Key!")
    print("   Dùng Private Key này để ký tin nhắn/claim airdrop sau này.")
    print("==================================================")

File: generate_did.js
const crypto = require('crypto');

// Bảng mã Base58btc
const ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

function base58Encode(buffer) {
    if (buffer.length === 0) return '';
    let digits = [0];
    for (let i = 0; i < buffer.length; i++) {
        for (let j = 0; j < digits.length; j++) digits[j] <<= 8;
        digits[0] += buffer[i];
        let carry = 0;
        for (let j = 0; j < digits.length; j++) {
            digits[j] += carry;
            carry = (digits[j] / 58) | 0;
            digits[j] %= 58;
        }
        while (carry > 0) {
            digits.push(carry % 58);
            carry = (carry / 58) | 0;
        }
    }
    for (let i = 0; i < buffer.length && buffer[i] === 0; i++) {
        digits.push(0);
    }
    return digits.reverse().map(d => ALPHABET[d]).join('');
}

function generateDidKey() {
    const { privateKey, publicKey } = crypto.generateKeyPairSync('ed25519');
    
    // Lấy 32 bytes raw public key từ DER SPKI
    const pubSpki = publicKey.export({ format: 'der', type: 'spki' });
    const pubBytes = pubSpki.subarray(pubSpki.length - 32);
    
    // Lấy 32 bytes raw private key từ DER PKCS8
    const privPkcs8 = privateKey.export({ format: 'der', type: 'pkcs8' });
    const privBytes = privPkcs8.subarray(privPkcs8.length - 32);
    
    // Multicodec prefix cho Ed25519 = 0xed01
    const multicodecPub = Buffer.concat([Buffer.from([0xed, 0x01]), pubBytes]);
    
    // Thêm tiền tố multibase 'z' cho base58btc
    const didKey = 'did:key:z' + base58Encode(multicodecPub);
    
    return {
        didKey,
        publicKeyHex: pubBytes.toString('hex'),
        privateKeyHex: privBytes.toString('hex')
    };
}

function signMessage(privateKeyHex, message) {
    const privBytes = Buffer.from(privateKeyHex, 'hex');
    // Đóng gói PKCS8 DER cho Node.js crypto
    const pkcs8Header = Buffer.from('302e020100300506032b657004220420', 'hex');
    const pkcs8Der = Buffer.concat([pkcs8Header, privBytes]);
    const keyObj = crypto.createPrivateKey({
        key: pkcs8Der,
        format: 'der',
        type: 'pkcs8'
    });
    
    const signature = crypto.sign(null, Buffer.from(message, 'utf-8'), keyObj);
    return signature.toString('hex');
}

if (require.main === module) {
    const keys = generateDidKey();
    console.log("==================================================");
    console.log("🔑 TECHNOCORE DID KEY GENERATOR (Node.js Zero-Dep)");
    print = console.log;
    print("==================================================");
    print(`DID Key         : ${keys.didKey}`);
    print(`Public Key (Hex): ${keys.publicKeyHex}`);
    print(`Private Key(Hex): ${keys.privateKeyHex}`);
    print("==================================================");
    print("⚠️  LƯU Ý BẢO MẬT: Giữ bí mật Private Key!");
    print("==================================================");
}

module.exports = { generateDidKey, signMessage };
